File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int>ans;
        int minus=0,count=0;
        int product=1, n=nums.size();
        for(int i=0;i<n;i++){
            if(nums[i]==0){
                count++;
                continue;
            }
            product*=nums[i];
        }
        for(int i=0;i<n;i++){
            if(nums[i]==0){
                if(count==1){
                    ans.push_back(product);
                }
                else{
                    ans.push_back(0);
                }
            }
            else{
                if(count>0){
                    ans.push_back(0);
                }
                else{
                    ans.push_back(product/nums[i]);
                }
            }
        }
        return ans;
    }
};

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        s = len(nums)
        preList = [1]*(s+1)
        sufList = [1]*(s+1)
        answer = [0]*s

        for i in range(0, s):
            preList[i+1] = preList[i]*nums[i]

            sufList[s-(i+1)] = sufList[s-(i)] * nums[s-(i+1)]

        # print(preList)
        # print(sufList)

        for i in range(0, s):
            answer[i] = preList[i]*sufList[i+1]

        return answer